File: Testing/www/js/testLinqGroupingOperators.js
/* global Bridge */

"use strict";

Bridge.define('ClientTestLibrary.Linq.AnagramEqualityComparer', {
    inherits: [Bridge.EqualityComparer$1(String)],
    equals: function (x, y) {
        return this.getCanonicalString(x) === this.getCanonicalString(y);
    },
    getHashCode: function (obj) {
        return Bridge.getHashCode(this.getCanonicalString(obj));
    },
    getCanonicalString: function (word) {
        if (word === null) {
            return null;
        }

        var wordChars = Bridge.String.toCharArray(word, 0, word.length);
        wordChars.sort();

        return String.fromCharCode.apply(null, wordChars);
    }
});

Bridge.define('ClientTestLibrary.Linq.TestLinqGroupingOperators', {
    statics: {
        test: function (assert) {
            assert.expect(3);

            // TEST
            var numbers = [2, 10, 3, 5, 30, 1, -15];
            var words = ["1.one", "3.three", "2.two", "22.twentytwo", "11.eleven", "30.thirty"];

            var numberGroups = (Bridge.Linq.Enumerable.from(numbers).groupBy(function (n) {
                return n % 5;
            }).select(function (g) {
                return { remainder: g.key(), numbers: g.toArray() };
            })).toArray();

            var numberGroupsExpected = [{ remainder: 2, numbers: [2] }, { remainder: 0, numbers: [10, 5, 30, -15] }, { remainder: 3, numbers: [3] }, { remainder: 1, numbers: [1] }];

            assert.deepEqual(numberGroups, numberGroupsExpected, "Group numbers by remainders");

            // TEST
            var wordGroups = (Bridge.Linq.Enumerable.from(words).groupBy(function (w) {
                return w.charCodeAt(0);
            }).select(function (g) {
                return { firstLetter: g.key(), words: g.toArray() };
            })).toArray();

            var wordGroupsExpected = [{ firstLetter: 49, words: ["1.one", "11.eleven"] }, { firstLetter: 51, words: ["3.three", "30.thirty"] }, { firstLetter: 50, words: ["2.two", "22.twentytwo"] }];

            assert.deepEqual(wordGroups, wordGroupsExpected, "Group words by first letters");

            // TEST
            var personGroups = (Bridge.Linq.Enumerable.from(ClientTestLibrary.Utilities.Person.getPersons()).groupBy(function (p) {
                return p.getGroup();
            }).select(function (g) {
                return { group: g.key(), persons: g.select(function (x) {
                    return x.getName();
                }).toArray() };
            })).toArray();

            var personGroupsExpected = [{ group: "A", persons: ["Frank"] }, { group: "C", persons: ["Zeppa", "Billy"] }, { group: "B", persons: ["John", "Dora", "Ian", "Mary"] }, { group: Bridge.cast(null, String), persons: ["Nemo"] }];

            assert.deepEqual(personGroups, personGroupsExpected, "Person group by Group field");
        },
        testComplexGrouping: function (assert) {
            assert.expect(1);

            // TEST
            var numbers = [2, 10, 3, 5, 30, 1, -15];
            var words = ["1.one", "3.three", "2.two", "22.twentytwo", "11.eleven", "30.thirty"];

            var complexGrouping = (Bridge.Linq.Enumerable.from(numbers).select(function (n) {
                return { number: n, words: (Bridge.Linq.Enumerable.from(words).where(function (w) {
                    return String.fromCharCode(w.charCodeAt(0)) === n.toString();
                }).groupBy(function (w) {
                    return w.charCodeAt(0);
                }).select(function (g) {
                    return { letter: g.key(), letterGroups: (Bridge.Linq.Enumerable.from(g).groupBy(function (l) {
                        return l;
                    }, function (l) {
                        return l;
                    }).select(function (mg) {
                        return { letter: mg.key(), letters: mg.toArray() };
                    })).toArray() };
                })).toArray() };
            })).toArray();

            var complexGroupingExpected = ClientTestLibrary.Linq.TestLinqGroupingOperators.getComplexGroupingExpectedResult();
            assert.deepEqual(complexGrouping, complexGroupingExpected, "Complex grouping for numbers and words");
        },
        testAnagrams: function (assert) {
            assert.expect(2);

            // TEST
            var anagrams = [" from ", " salt ", " earn ", " last ", " near ", " form "];

            var anagramsGroups = Bridge.Linq.Enumerable.from(anagrams).groupBy(function (w) {
                return w.trim();
            }, null, null, new ClientTestLibrary.Linq.AnagramEqualityComparer()).select(function (x) {
                return { key: x.key(), words: x.toArray() };
            }).toArray();

            var anagramsGroupsExpected = [{ key: "from", words: [" from ", " form "] }, { key: "salt", words: [" salt ", " last "] }, { key: "earn", words: [" earn ", " near "] }];

            assert.deepEqual(anagramsGroups, anagramsGroupsExpected, "Anagram grouping with equality comparer");

            // TEST
            var anagramsGroups1 = Bridge.Linq.Enumerable.from(anagrams).groupBy(function (w) {
                return w.trim();
            }, function (a) {
                return a.toUpperCase();
            }, null, new ClientTestLibrary.Linq.AnagramEqualityComparer()).select(function (x) {
                return { key: x.key(), words: x.toArray() };
            }).toArray();
            var anagramsGroupsExpected1 = [{ key: "from", words: [" FROM ", " FORM "] }, { key: "salt", words: [" SALT ", " LAST "] }, { key: "earn", words: [" EARN ", " NEAR "] }];

            assert.deepEqual(anagramsGroups1, anagramsGroupsExpected1, "Anagram grouping with equality compare and upper case");
        },
        getComplexGroupingExpectedResult: function () {
            var complexGroupingExpected = [{ number: 2, words: [{ letter: 50, letterGroups: [{ letter: "2.two", letters: ["2.two"] }, { letter: "22.twentytwo", letters: ["22.twentytwo"] }] }] }, { number: 10, words: [] }, { number: 3, words: [{ letter: 51, letterGroups: [{ letter: "3.three", letters: ["3.three"] }, { letter: "30.thirty", letters: ["30.thirty"] }] }] }, { number: 5, words: [] }, { number: 30, words: [] }, { number: 1, words: [{ letter: 49, letterGroups: [{ letter: "1.one", letters: ["1.one"] }, { letter: "11.eleven", letters: ["11.eleven"] }] }] }, { number: -15, words: [] }];

            return complexGroupingExpected;
        }
    }
});



Bridge.init();

File: Testing/tests/client/Linq/TestLinqGroupingOperators.cs
using Bridge;
using Bridge.QUnit;
using ClientTestLibrary.Utilities;
using System.Collections.Generic;
using System.Linq;

namespace ClientTestLibrary.Linq
{
    [FileName("testLinqGroupingOperators.js")]
    public class AnagramEqualityComparer : EqualityComparer<string>
    {
        public override bool Equals(string x, string y)
        {
            return GetCanonicalString(x) == GetCanonicalString(y);
        }

        public override int GetHashCode(string obj)
        {
            return GetCanonicalString(obj).GetHashCode();
        }

        private string GetCanonicalString(string word)
        {
            if (word == null)
            {
                return null;
            }

            var wordChars = word.ToCharArray();
            wordChars.JsSort();

            return new string(wordChars);
        }
    }

    internal class TestLinqGroupingOperators
    {
        public static void Test(Assert assert)
        {
            assert.Expect(3);

            // TEST
            var numbers = new[] { 2, 10, 3, 5, 30, 1, -15 };
            var words = new[] { "1.one", "3.three", "2.two", "22.twentytwo", "11.eleven", "30.thirty" };

            var numberGroups =
                    (from n in numbers
                     group n by n % 5 into g
                     select new
                     {
                         Remainder = g.Key,
                         Numbers = g.ToArray()
                     }).ToArray();

            var numberGroupsExpected = new[]
                {
                    new {Remainder = 2, Numbers = new[] { 2 } },
                    new {Remainder = 0, Numbers = new[] { 10, 5, 30, -15 } },
                    new {Remainder = 3, Numbers = new[] { 3 } },
                    new {Remainder = 1, Numbers = new[] { 1 } }
                };

            assert.DeepEqual(numberGroups, numberGroupsExpected, "Group numbers by remainders");

            // TEST
            var wordGroups =
                    (from w in words
                     group w by w[0] into g
                     select new
                     {
                         FirstLetter = g.Key,
                         Words = g.ToArray()
                     }).ToArray();

            var wordGroupsExpected = new[]
                {
                    new {FirstLetter = '1', Words = new[] { "1.one", "11.eleven" } },
                    new {FirstLetter = '3', Words = new[] { "3.three", "30.thirty" } },
                    new {FirstLetter = '2', Words = new[] { "2.two", "22.twentytwo" } }
                };

            assert.DeepEqual(wordGroups, wordGroupsExpected, "Group words by first letters");

            // TEST
            var personGroups =
                   (from p in Person.GetPersons()
                    group p by p.Group into g
                    select new
                    {
                        Group = g.Key,
                        Persons = g.Select(x => x.Name).ToArray()
                    }
                   ).ToArray();

            var personGroupsExpected = new object[]
            {
                new { Group = "A", Persons = new [] {"Frank"} },
                new { Group = "C", Persons = new [] {"Zeppa", "Billy"} },
                new { Group = "B", Persons = new [] {"John", "Dora", "Ian", "Mary"} },
                new { Group = (string)null, Persons = new [] {"Nemo"} }
            };

            assert.DeepEqual(personGroups, personGroupsExpected, "Person group by Group field");
        }

        public static void TestComplexGrouping(Assert assert)
        {
            assert.Expect(1);

            // TEST
            var numbers = new[] { 2, 10, 3, 5, 30, 1, -15 };
            var words = new[] { "1.one", "3.three", "2.two", "22.twentytwo", "11.eleven", "30.thirty" };

            var complexGrouping =
            (
               from n in numbers
               select
                   new
                   {
                       Number = n,
                       Words =
                           (
                           from w in words
                           where w[0].ToString() == n.ToString()
                           group w by w[0] into g
                           select
                               new
                               {
                                   Letter = g.Key,
                                   LetterGroups =
                                       (
                                       from l in g
                                       group l by l into mg
                                       select new
                                       {
                                           Letter = mg.Key,
                                           Letters = mg.ToArray()
                                       }
                                       ).ToArray()
                               }
                           ).ToArray()
                   }
           ).ToArray();

            var complexGroupingExpected = GetComplexGroupingExpectedResult();
            assert.DeepEqual(complexGrouping, complexGroupingExpected, "Complex grouping for numbers and words");
        }

        public static void TestAnagrams(Assert assert)
        {
            assert.Expect(2);

            // TEST
            var anagrams = new[]{
                    " from ",
                    " salt ",
                    " earn ",
                    " last ",
                    " near ",
                    " form "
            };

            var anagramsGroups = anagrams.GroupBy(w => w.Trim(), new AnagramEqualityComparer())
                                   .Select(x => new
                                   {
                                       Key = x.Key,
                                       Words = x.ToArray()
                                   })
                                   .ToArray();

            var anagramsGroupsExpected = new[]
                {
                    new {Key = "from",  Words = new []{ " from ", " form "} },
                    new {Key = "salt",  Words = new []{ " salt ", " last "} },
                    new {Key = "earn",  Words = new []{ " earn ", " near "} }
                };

            assert.DeepEqual(anagramsGroups, anagramsGroupsExpected, "Anagram grouping with equality comparer");

            // TEST
            var anagramsGroups1 = anagrams.GroupBy(w => w.Trim(), a => a.ToUpper(), new AnagramEqualityComparer())
                       .Select(x => new
                       {
                           Key = x.Key,
                           Words = x.ToArray()
                       })
                       .ToArray();
            var anagramsGroupsExpected1 = new[]
                {
                    new {Key = "from",  Words = new []{ " FROM ", " FORM "} },
                    new {Key = "salt",  Words = new []{ " SALT ", " LAST "} },
                    new {Key = "earn",  Words = new []{ " EARN ", " NEAR "} }
                };

            assert.DeepEqual(anagramsGroups1, anagramsGroupsExpected1, "Anagram grouping with equality compare and upper case");
        }

        private static object GetComplexGroupingExpectedResult()
        {
            var complexGroupingExpected = new object[]
            {
                new
                {
                    Number = 2,
                    Words = new []
                    { new
                        {
                            Letter = '2',
                            LetterGroups = new[]
                            {
                                new { Letter = "2.two", Letters = new []{"2.two"} },
                                new { Letter = "22.twentytwo", Letters = new []{"22.twentytwo"} }
                            }
                        }
                    }
                },
                new
                {
                    Number = 10, Words = new object[] { }
                },
                new
                {
                    Number = 3,
                    Words = new []
                    {
                        new
                        {
                            Letter = '3',
                            LetterGroups = new[]
                            {
                                new { Letter = "3.three", Letters = new []{"3.three"} },
                                new { Letter = "30.thirty", Letters = new []{"30.thirty"} }
                            }
                        }
                    }
                },
                new
                {
                    Number = 5, Words = new object[] { }
                },
                new
                {
                    Number = 30, Words = new object[] { }
                },
                new
                {
                    Number = 1,
                    Words = new []
                    {
                        new
                        {
                            Letter = '1',
                            LetterGroups = new[]
                            {
                                new { Letter = "1.one", Letters = new []{"1.one"} },
                                new { Letter = "11.eleven", Letters = new []{"11.eleven"} }
                            }
                        }
                    }
                },
                new
                {
                    Number = -15, Words = new object[] { }
                },
            };

            return complexGroupingExpected;
        }
    }
}
